### PR TITLE
some microoptimizations for the resolver

### PR DESCRIPTION
**Pull request by KristofferC**
These are results for `Pkg.update` in an environment where I added a bunch of big packages (Plots, Makie, DifferentialEquations, ApproxFun etc).

Master:

```
ENV["JULIA_PKGRESOLVE_ACCURACY"] = 1
  966.591 ms (6843000 allocations: 1.01 GiB)

ENV["JULIA_PKGRESOLVE_ACCURACY"] = 2
  2.124 s (11577823 allocations: 3.01 GiB)
```

PR:

```
ENV["JULIA_PKGRESOLVE_ACCURACY"] = 1
  602.627 ms (5432062 allocations: 301.72 MiB)

ENV["JULIA_PKGRESOLVE_ACCURACY"] = 2
   1.308 s (7230777 allocations: 368.51 MiB)
```

The performance improvements here only matter when the greedy solver fails.

The reason for this is that https://github.com/JuliaLang/Pkg.jl/issues/1949 suggests that we might want to bump the resolver accuracy but it would be a bit sad if we lost performance from doing that. The optimized resolver is now pretty close for accuracy 2, as the previous one with accuracy 1.
 
cc @carlobaldassi

### CODE REVIEW

File: src/Resolve/maxsum.jl [hunk 1]
```diff
@@ -346,6 +358,10 @@ function try_simplify_graph_soft!(graph, sources)
     return true
 end
 
+struct Unsat
+    p0::Int
```
Comment (KristofferC) [inline src/Resolve/maxsum.jl:362]: This was mostly to avoid using `try catch` as control flow.

**Comment by carlobaldassi** [comment]
This all looks good to me, thanks.
As for increasing the default accuracy, it might be that with the growing ecosystem and larger and more complicated dependency graphs it's going to be necessary. I should think if there is any way we can come up with a better heuristic to set it, but at the moment nothing particularly clever comes to mind.

File: src/Resolve/maxsum.jl [hunk 2]
```diff
@@ -235,6 +244,9 @@ function iterate!(graph::Graph, msgs::Messages, perm::NodePerm)
     for p0 in perm
         graph.ignored[p0] && continue
         maxdiff0 = update!(p0, graph, msgs)
+        if maxdiff0 isa Unsat
+            return Unsat
```
Comment (carlobaldassi) [inline src/Resolve/maxsum.jl:248]: Shouldn't this return `maxdiff0`, so as to include the same `p0` field inside the `Unsat` object?

Comment (KristofferC) [reply src/Resolve/maxsum.jl:248]: eeh, yes. Seems we are missing some test coverage here..

**Pull request merged**